File: Code Generation/responses/cpp/gemini/medium/sum-of-k-subarrays-with-length-at-least-m.cpp
#include <vector>
#include <numeric>
#include <algorithm>

using namespace std;

class Solution {
public:
    long long maxSum(vector<int>& nums, int k, int m) {
        int n = nums.size();
        
        // Prefix sums to calculate subarray sums in O(1)
        vector<long long> prefix_sum(n + 1, 0);
        for (int i = 0; i < n; ++i) {
            prefix_sum[i + 1] = prefix_sum[i] + nums[i];
        }

        // dp[i][j] is the maximum sum using i subarrays from the first j elements.
        // We use a 1D DP or two rows to save space, but given N=2000, 
        // a 2D array or two rows of size N+1 is fine.
        const long long INF = 1e18;
        vector<long long> dp(n + 1, -INF);
        
        // Base case: 0 subarrays result in a sum of 0
        for (int j = 0; j <= n; ++j) {
            dp[j] = 0;
        }

        for (int i = 1; i <= k; ++i) {
            vector<long long> next_dp(n + 1, -INF);
            long long max_prev_dp = -INF;
            
            // To pick the i-th subarray ending at index j-1 (1-based index j),
            // the subarray must start at some index 'start' such that j - start >= m.
            // So start <= j - m.
            // We need to look at dp[start] (from previous i-1 subarrays) 
            // plus the sum of nums[start...j-1].
            
            for (int j = m * i; j <= n; ++j) {
                // Option 1: The i-th subarray ends exactly at j-1 and has length m.
                // We transition from the best solution of i-1 subarrays ending at or before j-m.
                // However, we can extend a previously started i-th subarray.
                
                // To optimize:
                // Let f[i][j] be max sum of i subarrays within first j elements.
                // f[i][j] = max(f[i][j-1], max_{p <= j-m} (f[i-1][p] + sum(p...j-1)))
                
                // Let's redefine:
                // dp[j] = max sum of i-1 subarrays using first j elements.
                // next_dp[j] = max sum of i subarrays using first j elements.
            }
            
            // Correct approach for i-th subarray:
            // best_prev captures the max of (dp[p] - prefix_sum[p]) for p <= j - m
            long long best_prev = -INF;
            for (int j = 1; j <= n; ++j) {
                // We can potentially start a new m-length subarray ending at j
                if (j >= m) {
                    best_prev = max(best_prev, dp[j - m] - prefix_sum[j - m]);
                }
                
                next_dp[j] = next_dp[j - 1]; // Case: i-th subarray ends before index j-1
                if (best_prev != -INF) {
                    next_dp[j] = max(next_dp[j], best_prev + prefix_sum[j]);
                }
            }
            dp = move(next_dp);
        }

        return dp[n];
    }
};

File: Code Generation/responses/python/gemini/medium/sum-of-k-subarrays-with-length-at-least-m.py
from typing import List

class Solution:
    def maxSum(self, nums: List[int], k: int, m: int) -> int:
        n = len(nums)
        
        S = [0] * (n + 1)
        for i in range(n):
            S[i + 1] = S[i] + nums[i]
            
        prev_dp = [0] * (n + 1)
        
        for j in range(1, k + 1):
            new_dp = [float('-inf')] * (n + 1)
            max_val = float('-inf')
            
            for i in range(j * m, n + 1):
                p = i - m
                
                if prev_dp[p] != float('-inf'):
                    val = prev_dp[p] - S[p]
                    if val > max_val:
                        max_val = val
                        
                new_dp[i] = new_dp[i - 1]
                
                if max_val != float('-inf'):
                    current_sum = S[i] + max_val
                    if current_sum > new_dp[i]:
                        new_dp[i] = current_sum
                        
            prev_dp = new_dp
            
        return int(prev_dp[n])